data/policies/call_casting.cas

```diff
@@ -33,9 +33,30 @@ resource xyz inherits abc {
 	}
 }
 
-// TODOs once we add the ability to cast to
-// non-inherited things
-//
-//domain asd {
-//	asd<abc>.read(this);
-//}
+domain asd {
+	asd<abc>.read(this);
+	foo<abc>.read(this);
+}
+
+domain jkl inherits boo {
+	jkl<boo>.read_boo_tmp(this);
+}
+
+virtual resource tmp {
+	@associated_call
+	fn associated_call_from_tmp(domain source) {
+		allow(source, tmp, file, [read]);
+	}
+}
+
+@associate([tmp])
+virtual domain boo {
+	// Creates new resources boo.tmp and implicitly calls
+	// boo.tmp.associated_call_from_tmp(boo)
+	//
+	// boo.tmp inherits tmp
+
+	fn read_boo_tmp(domain source) {
+		allow(source, boo.tmp, file, [read]);
```

Comment:
Does the "this.tmp" case work as well?

I'd think if this line said "this.tmp" instead of "boo.tmp" that we'd expect "this" to still resolve to jkl, despite the cast, which only makes it call boo's function and pass type validation, but still is "really" jkl under the hood.  Is that an accurate understanding of the intent of the code?  And can we add validation to ensure it's the case?

Reply:
That is the case, and I was able to validate.

Reply:
It looks like the `this.tmp` is resolved when the macro gets written out to cil.  So by adding:
`jkl<boo>.read_boo_this_tmp(this);` &
```
fn read_boo_this_tmp(domain source) {
    allow(source, this.tmp, file, [read]);
}
```

I get the following cil lines:
`(macro boo-read_boo_this_tmp ((type this) (type source)) (allow source boo-tmp (file (read))))` & `(call boo-read_boo_this_tmp (jkl jkl))`

Reply:
Ah, yeah, that makes sense that that's how it's working now.  I don't think that's the behavior we want though.  I think the expectation is that when I run `jkl<boo>.read_boo_tmp(this)`, I get the access on `jkl.tmp` rather than `boo.tmp`.

The problem is how to achieve that.  The overwrite `this` method works for casting `jkl`, but I don't think we have an immediately obvious way to also convert `boo-tmp` into `jkl-tmp` as well.  I can't immediately think of a way to handle this besides generating a new macro that matches the current one except changing `boo-tmp` into `jkl-tmp`, which isn't the most desirable thing in my mind.  Any ideas of solutions here I'm missing?

Reply:
Unless cil deals with `this` in a sort of find replace kind of way, I agree we would need a new macro for every cast of `read_boo_this_tmp`.  It's a bit out of scope for this work, but if you point me to where we resolve the macros, I can try to find a fix.

Reply:
https://github.com/dburgener/cascade/issues/144